derivation-tests/shared/src/test/scala/zio/prelude/derivation/tests/DeriveDebugSpec.scala

```diff
@@ -0,0 +1,59 @@
+package zio.prelude.derivation.tests
+
+import zio._
+import zio.prelude._
+import zio.prelude.derivation.DeriveDebug
+import zio.test._
+
+object DeriveDebugSpec extends ZIOSpecDefault {
+  sealed trait Entity
+  object Entity {
+    case class Person(name: String, address: Address) extends Entity
+    object Person {
+      implicit val debug: Debug[Person] = DeriveDebug.gen[Person]
+    }
+
+    case class Organization(name: String, contacts: List[Person]) extends Entity
+
+    // need to keep it as not implicit, or the derivation would reference itself for the subtype's instances
+    // and runs into StackOverflow in the runtime
+    val debug: Debug[Entity] = DeriveDebug.gen[Entity]
```

Comment:
This is pretty terrible. I think if you use `-Xsource:3` it will fix this.

Reply:
`-Xsource:3` can't fix this either. :-(

The underlying problem of this can be implicit is because `Debug` is contravariant, and `Magnolia`/`zio-deriving` would resolve the subtypes' typeclass instance to this exact debug instance, instead of generating a new one.

When I add the `implicit` keyword under the magnolia implementation, the compiler would complain on this line that:
```
Implicit resolves to enclosing value debug
[warn]     implicit val debug: Debug[Entity] = DeriveDebug.gen[Entity]
```

And during runtime, as it always refer to itself, we will get a StackOverflowException in both `Magnolia` and `zio-deriving`.

But if I remove the `implicit` keyword, both libraries can generate a dedicated instance for a sealed trait's subtypes. 

A workaround I can think of is to mimic `DeriveGen` from `zio-test`, making `DeriveDebug` a invariant typeclass, which can provide the `Debug` instance (like the invariant `DeriveGen` can provide a instance for `Gen[-R, +A]`). So that the client side can always declare the `Debug` instance `implicit`.

How do you think about this solution?